compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -111,6 +140,85 @@ impl OptimizationFinder<'b, 'tcx> {
         }
     }
 
+    fn find_eq_not(&mut self, rvalue: &Rvalue<'tcx>, location: Location) -> Option<()> {
+        // Optimize the sequence
+        // _4 = Eq(move _5, const 2_u8);
+        // StorageDead(_5);
+        // _3 = Not(move _4);
+        //
+        // into _3 = Ne(move _5, const 2_u8)
+        if let Rvalue::UnaryOp(UnOp::Not, op) = rvalue {
+            let place = op.place()?;
+            // See if we can find a Eq that assigns `place`.
+            // We limit the search to 3 statements lookback.
+            // Usually the first 2 statements are `StorageDead`s for operands for Eq.
+            // We record what is marked dead so that we can reorder StorageDead so it comes after Ne
+
+            // We will maximum see 2 StorageDeads
+            let mut seen_storage_deads: SmallVec<[_; 2]> = SmallVec::new();
+            let lower_index = location.statement_index.saturating_sub(3);
+            for (stmt_idx, stmt) in self.body.basic_blocks()[location.block].statements
+                [lower_index..location.statement_index]
+                .iter()
+                .enumerate()
+                .rev()
+            {
+                match &stmt.kind {
+                    rustc_middle::mir::StatementKind::Assign(box (l, r)) => {
+                        if *l == place {
+                            match r {
+                                Rvalue::BinaryOp(BinOp::Eq, op1, op2) => {
+                                    // We need to make sure that the StorageDeads we saw are for
+                                    // either `op1`or `op2` of Eq. Else we bail the optimization.
+                                    for (dead_local, _) in seen_storage_deads.iter() {
+                                        let dead_local_matches = [op1, op2].iter().any(|x| {
+                                            Some(*dead_local) == x.place().map(|x| x.local)
+                                        });
+                                        if !dead_local_matches {
+                                            return None;
+                                        }
+                                    }
+
+                                    // We need to make sure that the Ne we are going to insert comes before the
+                                    // StorageDeads so we want to swap the StorageDead closest to Eq with Ne.
```

Comment:
I thought we are replacing the comparison, not the `StorageDead`?

Reply:
I think the confusion comes from the use of "insert"? We are replacing the Not with a Ne. And since we are going to use the operands from Eq in Ne, we need to make sure that we do not mark these operands as dead before they are used. I'll push up a new commit where I try to explain this better.